### PR TITLE
Introduce in-Segment Trim for GroupBy OrderBy Query

### PR DESCRIPTION
**Pull request by wuwenw**
## Description

One of the major bottlenecks for the current GroupBy OrderBy query on high cardinality columns is the merge phase. Essentially every segment brings a large number of intermediate results to a global concurrent map for further aggregation and merge, which takes up a lot of space and is very time-consuming. This PR introduces an optimization option that each segment trims its intermediate results to a given size. The size is configurable by the user and is guaranteed to be max(limit N * 5, 5000). It won't affect accuracy much but reduces the running time for high cardinality dataset. ~5 times faster for String data with 10M cardinality. This option is turned off by default to ensure backward compatibility. 
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [x] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes

Optimized GroupBy OrderBy queries by introducing an in-segment trim option that can significantly reduce the size of intermediate results and speed up the execution.

## Documentation

GroupBy OrderBy In-Segment Trim
This option is used to apply an additional filter in the segments for the GroupBy OrderBy query. Based on a configurable trim_size k, each segment will sort and send back the top k results only. It will slightly decrease the accuracy but boost up the performance for high-cardinality data.

Related query options: 
SEGMENT TRIM ON
SEGMENT TRIM SIZE

Related server config:
enable.segment.group.trim
size.segment.group.trim

Related query keywords:
LIMIT N

SEGMENT TRIM ON and enable.segment.group.trim are used to enable/disable this feature. By default, it is turned off and segments will send back all results. 

If it is enabled, then actual trim_size is calculated based on the parameters (SEGMENT TRIM SIZE and size.segment.group.trim) and query keyword LIMIT N.

Trim_size = max(5*LIMIT, minTrimSize) where minTrimSize is set up by the user (Default value: 5000)

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 1]
```diff
@@ -61,7 +61,11 @@
   public static final String MAX_INITIAL_RESULT_HOLDER_CAPACITY_KEY = "max.init.group.holder.capacity";
   public static final int DEFAULT_MAX_INITIAL_RESULT_HOLDER_CAPACITY = 10_000;
   public static final String NUM_GROUPS_LIMIT = "num.groups.limit";
+  public static final String SEGMENT_TRIM_SIZE = "num.segment.trim.limit";
+  public static final String SEGMENT_TRIM_OPT = "bool.segment.trim";
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:65]:
```suggestion
  public static final String ENABLE_SEGMENT_GROUP_TRIM = "enable.segment.group.trim";
```

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 2]
```diff
@@ -61,7 +61,11 @@
   public static final String MAX_INITIAL_RESULT_HOLDER_CAPACITY_KEY = "max.init.group.holder.capacity";
   public static final int DEFAULT_MAX_INITIAL_RESULT_HOLDER_CAPACITY = 10_000;
   public static final String NUM_GROUPS_LIMIT = "num.groups.limit";
+  public static final String SEGMENT_TRIM_SIZE = "num.segment.trim.limit";
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:64]: Let's not make this configurable in step 1. Use `max(limit * 5, 5000)` (same as inter segment trim). You can use `GroupByUtils` to get the trim size from the `limit`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java [hunk 3]
```diff
@@ -117,6 +119,19 @@ public IntermediateResultsBlock(AggregationFunction[] aggregationFunctions,
     _dataSchema = dataSchema;
   }
 
+  /**
+   * Constructor for aggregation group-by order-by result with {@link AggregationGroupByResult}.
+   */
+  public IntermediateResultsBlock(AggregationFunction[] aggregationFunctions,
+                                  @Nullable AggregationGroupByResult aggregationGroupByResults,
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java:126]: Follow https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup to setup the code style in your IDE

File: pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java [hunk 4]
```diff
@@ -357,4 +372,88 @@ public Comparable extract(Record record) {
       }
     }
   }
+
+  public static class fullIntermediateResult extends IntermediateRecord {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java:376]: You can directly change the `IntermediateRecord` to include the record

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java:376]: Class name should be capitalized

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java [hunk 5]
```diff
@@ -144,4 +147,22 @@ protected void aggregate(TransformBlock transformBlock, int length, int function
...
+    // Check if a trim is necessary
+    int keyNum = 0;
+    if (_hasMVGroupByExpression) {
+      keyNum = _mvGroupKeys.length;
+    } else {
+      keyNum = _svGroupKeys.length;
+    }
+    Iterator<GroupKeyGenerator.GroupKey> groupKeyIterator = _groupKeyGenerator.getGroupKeys();
+    if (keyNum > threshold && threshold != -1) {
+      return tableResizer.trimInSegmentResults(groupKeyIterator, _groupByResultHolders, threshold);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java:162]: threshold should be calculated based on the `limit` instead of hardcoded

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:64]: Not making it configurable will change the default behavior?

**Review by siddharthteotia — COMMENTED** [review]
@wuwenw thanks for the PR. Yet to fully go through it. Please give a day to go through it.

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:64]: We want to configure whether to enable it, but not the threshold in step 1. If it is off, then it is the same as the current behavior

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java (hunk 5 above)
Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java:163]: Now the threshold is calculated in the instance plan maker

> threshold should be calculated based on the `limit` instead of hardcoded

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java (hunk 3 above)
Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java]: Followed

File: pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java (hunk 4 above)
Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java]: Now directly extend IntermediateRecord

> You can directly change the `IntermediateRecord` to include the record

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java (hunk 1 above)
Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java]: Changed

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 6]
```diff
@@ -51,40 +51,56 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(InstancePlanMakerImplV2.class);
   private final int _maxInitialResultHolderCapacity;
   // Limit on number of groups stored for each segment, beyond which no new group will be created
   private final int _numGroupsLimit;
   // Used for SQL GROUP BY (server combine)
   private final int _groupByTrimThreshold;
+  private final boolean _enableSegmentGroupTrim;
+
+  @VisibleForTesting
+  private int _inSegmentTrimLimit = -1;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:79]: Since we need it for testing purpose, let's directly make it configurable as `_minSegmentTrimSize` and default to 5000

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 7]
```diff
@@ -105,8 +121,61 @@ public InstancePlanMakerImplV2(QueryExecutorConfig queryExecutorConfig) {
...
+        _maxInitialResultHolderCapacity, _numGroupsLimit, _enableSegmentGroupTrim);
+  }
+
+  /**
+   * Returns {@code true} if the given aggregation-only without filter QueryContext can be solved with segment metadata,
+   * {@code false} otherwise.
+   * <p>Aggregations supported: COUNT
+   */
+  @VisibleForTesting
+  static boolean isFitForMetadataBasedPlan(QueryContext queryContext) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:137]: Why moving these 2 methods? Seems unrelated

Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:144]: These 2 methods are moved by the format-tools

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java [hunk 8]
```diff
@@ -144,4 +147,22 @@ protected void aggregate(TransformBlock transformBlock, int length, int function
   public AggregationGroupByResult getResult() {
     return new AggregationGroupByResult(_groupKeyGenerator, _aggregationFunctions, _groupByResultHolders);
   }
+
+  @Override
+  public Collection<TableResizer.IntermediateRecord> trimGroupByResult(boolean enableSegmentGroupTrim,
```
[the remark this answers is not in the document]
Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java]: > Remove `enableSegmentGroupTrim`. When this method is invoked, it should always trim the result

I believe a simple option to turn on/off is good for users

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java [hunk 9]
```diff
@@ -281,6 +294,14 @@ public void setNumGroupsLimitReached(boolean numGroupsLimitReached) {
     _numGroupsLimitReached = numGroupsLimitReached;
   }
 
+  /**
+   * Get the collection of intermediate records
+   */
+  @Nullable
+  public Collection<IntermediateRecord> getIntermediateResult() {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java:301]:
```suggestion
  public Collection<IntermediateRecord> getIntermediateRecords() {
```

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 10]
```diff
@@ -52,39 +52,56 @@
...
-
   public static final String MAX_INITIAL_RESULT_HOLDER_CAPACITY_KEY = "max.init.group.holder.capacity";
   public static final int DEFAULT_MAX_INITIAL_RESULT_HOLDER_CAPACITY = 10_000;
   public static final String NUM_GROUPS_LIMIT = "num.groups.limit";
   public static final int DEFAULT_NUM_GROUPS_LIMIT = 100_000;
-
+  public static final String ENABLE_SEGMENT_GROUP_TRIM = "enable.segment.group.trim";
+  public static final boolean DEFAULT_ENABLE_SEGMENT_GROUP_TRIM = false;
+  public static final String SIZE_SEGMENT_GROUP_TRIM = "size.segment.group.trim";
+  public static final int DEFAULT_SEGMENT_TRIM_SIZE = -1;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:67]:
```suggestion
  public static final String MIN_SEGMENT_GROUP_TRIM_SIZE = "min.segment.group.trim.size";
  public static final int DEFAULT_MIN_SEGMENT_GROUP_TRIM_SIZE = -1;
```

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 11]
```diff
@@ -52,39 +52,56 @@
...
   public static final int DEFAULT_GROUPBY_TRIM_THRESHOLD = 1_000_000;
-
+  private static final Logger LOGGER = LoggerFactory.getLogger(InstancePlanMakerImplV2.class);
   private final int _maxInitialResultHolderCapacity;
   // Limit on number of groups stored for each segment, beyond which no new group will be created
   private final int _numGroupsLimit;
   // Used for SQL GROUP BY (server combine)
   private final int _groupByTrimThreshold;
+  private final boolean _enableSegmentGroupTrim;
+  private final int _minSegmentTrimSize;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:78]:
```suggestion
  private final int _minSegmentGroupTrimSize;
```

File: pinot-core/src/main/java/org/apache/pinot/core/util/GroupByUtils.java [hunk 12]
```diff
@@ -36,6 +36,15 @@ public static int getTableCapacity(int limit) {
     return Math.max(limit * 5, NUM_RESULTS_LOWER_LIMIT);
   }
 
+  /**
+   * Returns the capacity of the table required by the given query.
+   * NOTE: It returns {@code max(limit * 5, resultLowerLimit)} where resultLowerLimit is configured by the user
+   *      (Default: 5000)
+   */
+  public static int getTableCapacity(int limit, int resultLowerLimit) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/util/GroupByUtils.java:44]:
```suggestion
  public static int getTableCapacity(int limit, int minNumGroups) {
```

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/util/GroupByUtils.java:44]: Also change `NUM_RESULTS_LOWER_LIMIT` to public `DEFAULT_MIN_NUM_GROUPS` which can be accessed by the plan maker

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java (hunk 11 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:78]: As discussed offline, we can remove `_enableSegmentGroupTrim` and only maintain the trim size here

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/AggregationGroupByOrderByOperator.java [hunk 13]
```diff
@@ -106,8 +116,14 @@ protected IntermediateResultsBlock getNextBlock() {
       groupByExecutor.process(transformBlock);
     }
 
-    // Build intermediate result block based on aggregation group-by result from the executor
-    return new IntermediateResultsBlock(_aggregationFunctions, groupByExecutor.getResult(), _dataSchema);
+    // Trim is off or no need to trim
+    if (_trimSize == TRIM_OFF || groupByExecutor.getResultNum() <= _trimSize) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/AggregationGroupByOrderByOperator.java:120]:
```suggestion
    if (_trimSize <= 0 || groupByExecutor.getNumGroups() <= _trimSize) {
```

File: pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java [hunk 14]
```diff
@@ -65,6 +76,33 @@ public void setUp() {
...
+    for (int j = 0; j < _groupKeys.size(); ++j) {
+      _groupByResultHolders[0]
+          .setValueForKey(_groupKeys.get(j)._groupId, 10 % ((Double) _groupKeys.get(j)._keys[2]));
+      _groupByResultHolders[1]
+          .setValueForKey(_groupKeys.get(j)._groupId, 100 % ((Double) _groupKeys.get(j)._keys[2]));
+      _groupByResultHolders[2].setValueForKey(_groupKeys.get(j)._groupId, new IntOpenHashSet(new int[]{j}));
+      _groupByResultHolders[3].setValueForKey(_groupKeys.get(j)._groupId, new AvgPair(10, j + 1));
+    }
+
     //@formatter:on
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java:106]: Move this line to line 79

File: pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java [hunk 15]
```diff
@@ -43,10 +49,15 @@
       new DataSchema(new String[]{"d1", "d2", "d3", "sum(m1)", "max(m2)", "distinctcount(m3)", "avg(m4)"},
           new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.INT, DataSchema.ColumnDataType.DOUBLE, DataSchema.ColumnDataType.DOUBLE, DataSchema.ColumnDataType.DOUBLE, DataSchema.ColumnDataType.OBJECT, DataSchema.ColumnDataType.OBJECT});
   private static final int TRIM_TO_SIZE = 3;
+  private static final int NUM_RESULT_HOLDER = 4;
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java:52]: We should be able to use the same `TRIM_TO_SIZE` for inner segment trim

File: pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java [hunk 16]
```diff
@@ -65,6 +76,33 @@ public void setUp() {
         new Key(new Object[]{"c", 50, 4.0}),
         new Key(new Object[]{"c", 300, 5.0})
     );
+
+    // GroupKeys: d1: a0 ~ a14, d2: 10.0 ~ 24.0, d3: 1.0 ~ 15.0
+    _groupKeys = new LinkedList<>();
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java:81]: Let's make `_groupKeys` and `_groupByResultHolders` consistent with `_records` and `_keys` (5 groups, values in holder the same as in records)

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java [hunk 17]
```diff
@@ -41,6 +42,43 @@ public void testGroupByOrderByMVSQLResults(String query, List<Object[]> expected
...
   }
 
+  /**
+   * Tests the in-segment trim option for GroupBy OrderBy query
+   */
+  @Test(dataProvider = "orderByDataProvider")
+  public void testGroupByOrderByMVTrimOptLowLimitSQLResults(String query, List<Object[]> expectedResults,
+      long expectedNumEntriesScannedPostFilter, DataSchema expectedDataSchema) {
+
+    expectedResults = expectedResults.subList(0, expectedResults.size() / 2);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java:52]: I don't see the value of this test. This test only changes the limit of the query

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java [hunk 18]
```diff
@@ -41,6 +42,43 @@ public void testGroupByOrderByMVSQLResults(String query, List<Object[]> expected
...
+            expectedResults, expectedResults.size(), expectedDataSchema);
+  }
+
+  /**
+   * Tests the in-segment build option for GroupBy OrderBy query. (No trim)
+   */
+  @Test(dataProvider = "orderByDataProvider")
+  public void testGroupByOrderByMVTrimOptHighLimitSQLResults(String query, List<Object[]> expectedResults,
+      long expectedNumEntriesScannedPostFilter, DataSchema expectedDataSchema) {
+    InstancePlanMakerImplV2 planMaker = new InstancePlanMakerImplV2(expectedResults.size() + 1, true);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java:75]: Here we can simply put a very low `minSegmentGroupTrimSize` and expect the same result
```suggestion
    InstancePlanMakerImplV2 planMaker = new InstancePlanMakerImplV2(1, true);
```

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java [hunk 19]
```diff
@@ -41,6 +42,43 @@ public void testGroupByOrderByMVSQLResults(String query, List<Object[]> expected
...
+    QueriesTestUtils
+        .testInterSegmentResultTable(brokerResponse, 400000L, 0, expectedNumEntriesScannedPostFilter, 400000L,
+            expectedResults, expectedResults.size(), expectedDataSchema);
+  }
+
+  /**
+   * Tests the in-segment build option for GroupBy OrderBy query. (No trim)
+   */
+  @Test(dataProvider = "orderByDataProvider")
+  public void testGroupByOrderByMVTrimOptHighLimitSQLResults(String query, List<Object[]> expectedResults,
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java:73]:
```suggestion
  public void testGroupByOrderByMVSegmentTrimSQLResults(String query, List<Object[]> expectedResults,
```

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java [hunk 20]
```diff
@@ -72,6 +73,46 @@ public void testGroupByOrderByPQLResponse(String query, List<String[]> expectedG
...
   }
 
+  /**
+   * Tests the in-segment trim option for GroupBy OrderBy query
+   */
+  @Test(dataProvider = "orderBySQLResultTableProvider")
+  public void testGroupByOrderByTrimOptSQLLowLimitResponse(String query, List<Object[]> expectedResults,
+      long expectedNumDocsScanned, long expectedNumEntriesScannedInFilter, long expectedNumEntriesScannedPostFilter,
+      long expectedNumTotalDocs, DataSchema expectedDataSchema) {
+    expectedResults = expectedResults.subList(0, expectedResults.size() / 2);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:83]: Same as the comments in `InterSegmentOrderByMultiValueQueriesTest`

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java [hunk 21]
```diff
@@ -72,6 +73,46 @@ public void testGroupByOrderByPQLResponse(String query, List<String[]> expectedG
...
+        .testInterSegmentResultTable(brokerResponse, expectedNumDocsScanned, expectedNumEntriesScannedInFilter,
+            expectedNumEntriesScannedPostFilter, expectedNumTotalDocs, expectedResults, expectedResults.size(),
+            expectedDataSchema);
+  }
+
+  /**
+   * Tests the in-segment build option for GroupBy OrderBy query. (No trim)
+   */
+  @Test(dataProvider = "orderBySQLResultTableProvider")
+  public void testGroupByOrderByTrimOptHighLimitSQLResponse(String query, List<Object[]> expectedResults,
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:105]: Same as the comments in `InterSegmentOrderByMultiValueQueriesTest`

File: pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java (hunk 16 above)
Comment (wuwenw) [reply pinot-core/src/test/java/org/apache/pinot/core/data/table/TableResizerTest.java:86]: Changed _groupKeys and _groupResultholders to align with the old tests

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java (hunk 17 above)
Comment (wuwenw) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java:52]: This test aims to trigger the in-segment trim by providing a low limit number and a low trimSize, so the size of intermediateResult is greater than max(5*limit, trimSize). intermediateResult will be trimmed by the in-segment trim feature and then a limit will be applied (this is unavoidable). Still, we can test to see if everything works in the in-segment trim phase

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationGroupByOrderByPlanNode.java [hunk 22]
```diff
@@ -58,6 +63,14 @@ public AggregationGroupByOrderByPlanNode(IndexSegment indexSegment, QueryContext
     List<ExpressionContext> groupByExpressions = queryContext.getGroupByExpressions();
     assert groupByExpressions != null;
     _groupByExpressions = groupByExpressions.toArray(new ExpressionContext[0]);
+    _queryContext = queryContext;
+
+    // Only trim if there is OrderBy and has a positive trim size
+    if (queryContext.getOrderByExpressions() != null && minSegmentTrimSize > 0) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationGroupByOrderByPlanNode.java:69]: I feel we might want to pass `minSegmentTrimSize` to the operator, and the operator can decide whether to trim it or not

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 23]
```diff
@@ -187,6 +216,7 @@ public PlanNode makeStreamingSegmentPlanNode(IndexSegment indexSegment, QueryCon
     }
   }
 
+
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:219]: (nit) remove

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Please revise the tests as discussed offline

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationGroupByOrderByPlanNode.java (hunk 22 above)
Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationGroupByOrderByPlanNode.java:69]: Any reason for this behavior rather than figuring out in the plan?

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationGroupByOrderByPlanNode.java:69]: Because we have to do the actual comparison in the operator (during query run time), so imo more clear if we group the comparison in one place

Comment (wuwenw) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationGroupByOrderByPlanNode.java:69]: I see

File: pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java [hunk 24]
```diff
@@ -52,39 +53,53 @@
 import org.slf4j.LoggerFactory;
 
 
+
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:56]: (nit) remove

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise
